### PR TITLE
Murisi/fix masp vp mint case

### PR DESCRIPTION
**Pull request by murisi**
## Describe your changes
Addressed some of the MASP-IBC issues brought up during an audit. More specifically, the following changes have been made:
* The IBC internal address pre-balance is now only increased when a IbcMsgReceive packet requiring minting occurs
* Modified the MASP-VP to make an early exit if non-allowed keys are modified
* Updated/corrected outdated comments

## Checklist before merging 
- [x] If this PR has some consensus breaking changes, I added the corresponding `breaking::` labels
    - This will require 2 reviewers to approve the changes

### CODE REVIEW

File: crates/ibc/src/trace.rs [hunk 1]
```diff
@@ -108,13 +108,24 @@ pub fn is_nft_trace(
     }
 }
 
-/// Return true if the source of the given IBC trace is this chain
-pub fn is_sender_chain_source(
+/// Returns true if the denomination originally came from the sender chain, and
+/// false otherwise.
+pub fn is_sender_chain_source_str(
```
Comment (yito88) [inline crates/ibc/src/trace.rs:113]: just curious, why did you add `_str`?

Comment (murisi) [reply crates/ibc/src/trace.rs:113]: I added `_str` in order to distinguish https://github.com/anoma/namada/blob/763b5fc0b251144271d6d6ccf134aab76c91eacb/crates/ibc/src/lib.rs#L825 from https://github.com/anoma/namada/blob/763b5fc0b251144271d6d6ccf134aab76c91eacb/crates/ibc/src/lib.rs#L524 since they are both used in the same file. That being said, I should have just used import `as`.

Comment (murisi) [reply crates/ibc/src/trace.rs]: I've now reverted the naming and use import `as`.

File: crates/shielded_token/src/vp.rs [hunk 2]
```diff
@@ -869,7 +869,7 @@ fn apply_balance_component(
     })
 }
 
-// Verify that the pre balance + the Sapling value balance = the post balance
+// Verify that the pre balance - the Sapling value balance = the post balance
```
Comment (cwgoes) [inline crates/shielded_token/src/vp.rs:872]: Was logic actually changed here, or was this comment just wrong before?

**Review by cwgoes — CHANGES_REQUESTED** [review]
One question. I also think we should probably write a test-case for the revised functionality?

**Review by grarco — APPROVED** [review]
LGTM, thank you!

File: crates/shielded_token/src/vp.rs (hunk 2 above)
Comment (grarco) [reply crates/shielded_token/src/vp.rs:872]: Wrong comment but the logic is unchanged

**Review by cwgoes — APPROVED** [review]

**Review by yito88 — APPROVED** [review]
